mac/mcompile/keymap.cpp

```diff
@@ -0,0 +1,451 @@
+    else return rgkey_ShiftState;
+  }
+
+/**
+ * @brief  check for correct input parameter that will later be used in UCKeyTranslate()
+ * @param  shiftstate the currently used shiftstate
+ * @param  keycode 		the code of the key in question
+ * @return true if all parameters are OK;
+ * 				 false if not
+ */
+bool ensureValidInputForKeyboardTranslation(int shiftstate, int keycode) {
+  if (!(std::find(std::begin(ss_mac), std::end(ss_mac), shiftstate) != std::end(ss_mac)))
+    return false;
+
+  if (keycode > keycode_max)
+    return false;
+
+return true;
+}
+
+/**
+ * @brief  create a 3D-Vector containing data of the US keyboard and the currently used (underlying) keyboard
+ *          all_vector [ US_Keyboard  ]
+ *                 		           [KeyCode_US        ]
+ *                 		           [Keyval unshifted  ]
+ *                 		           [Keyval shifted    ]
+ *                     [Underlying Kbd]
+ *                   		         [KeyCode_underlying]
+ *                   		         [Keyval unshifted  ]
+ *                   		         [Keyval shifted    ]
+ * @param[in,out] all_vector 			Vector that holds the data of the US keyboard as well as the currently used (underlying) keyboard
+ * @param         keyboard_layout pointer to currently used (underlying) keyboard layout
+ * @return 0 on success;
+ * 				 1 if data of US keyboard was not written;
+ * 			   2 if data of underlying keyboard was not written
+ */
+int mac_createOneVectorFromBothKeyboards(vec_dword_3D& all_vector, const UCKeyboardLayout* keyboard_layout) {
+  // store contents of the English (US) keyboard in all_vector
+  if (mac_write_US_ToVector(all_vector)) {
+    printf("ERROR: can't write full US to Vector \n");
+    return 1;
+  }
+
+  // add contents of underlying keyboard to all_vector
+  if (mac_append_underlying_ToVector(all_vector, keyboard_layout)) {
+    printf("ERROR: can't append underlying ToVector \n");
+    return 2;
+  }
+  return 0;
+}
+
+/**
+ * @brief  write data of the US keyboard into a 3D-Vector which later will contain
+ *         data of the US keyboard and the currently used (underlying) keyboard
+ * @param[in,out] vec_us Vector that holds the data of the US keyboard
+ * @return 0 on success;
+ * 				 1 if data of US keyboard was not written;
+ */
+int mac_write_US_ToVector(vec_dword_3D& vec_us) {
+  // Values for US        :     A,  S,  D,  F,  H,  G,  Z,  X, C,  V,  §, B,  Q,  W,  E,  R,  Y,  T, 1, 2, 3, 4, 6, 5, =, 9, 7, -, 8, 0,  ],  O,  U,  [,  I,  P,CR,  L,  J, ',  K, ;,  \, ,, /,  N,  M, .
+  std::vector<int> us_Base  = {97,115,100,102,104,103,122,120,99,118,167,98,113,119,101,114,121,116,49,50,51,52,54,53,61,57,55,45,56,48, 93,111,117, 91,105,112,13,108,106,39,107,59, 92,44,47,110,109,46};
+  std::vector<int> us_Shift = {65, 83, 68, 70, 72, 71, 90, 88,67, 86,177,66, 81, 87, 69, 82, 89, 84,33,64,35,36,94,37,43,40,38,95,42,41,125, 79, 85,123, 73, 80,13, 76, 74,34, 75,58,124,60,63, 78, 77,62};
```

Comment:
It'd be good for these constants to be defined outside the function (and as consts)